### ISSUE TITLE
rename MFVI as KLqp ?

### ISSUE THREAD

**Issue by akucukelbir** [blei-lab/edward#129]
MFVI is currently a catch-all class: it captures BBVI, ADVI (not implemented yet) and various extensions.

perhaps it makes sense to rename it `KLqp`?

or perhaps it makes sense to rename both KL-based methods as `KLforward` and `KLreverse`?

what do you guys think?

**Comment by dustinvtran** [comment]
Before deciding, perhaps it's useful to think where `MFVI` situates among all current and future implemented algorithms. Of course, we don't have the foresight to know all algorithms or even know the best conceptual foundation on how to organize them. But it could be a useful thought exercise.

Here is my attempt at a flattened list below (my cognitive biases already led me to put down _some_ structure as I wrote them down). Not all apply to the same class of probability models.

variational inference
+ coordinate ascent variational inference (conjugate exponential family updates, variational message passing, original SVI, structured SVI)
+ variational inference with score function gradient (mean-field, structured, hierarchical variational models)
+ variational inference with reparameterization gradient (mean-field, structured, hierarchical variational models)
+ KL(p || q)
+ D_f(p ||q) for tractable f's
+ expectation propagation (local KL(p || q), assumed density filtering)
+ power expectation propagation (local D_alpha(p || q), including black box alpha, general message passing)
+ memoized VI
+ streaming VB
+ population VI
+ collapsed VI
+ generative adversarial networks
+ wake sleep
+ reweighted wake sleep
+ importance weighted VI
+ maximum a posteriori

other algorithms/tasks

+ Laplace approximation
   + integrated nested Laplace approximation
+ marginal optimization
   + EM (exponential family, Baum-Welch for state space models)
   + Monte Carlo EM
   + variational EM
   + online/incremental EM
   + auxiliary variable EM, e.g., ECM
   + GMO
   + Bayesian risk minimization
+ Monte Carlo
   + MCMC (Metropolis-Hastings, Gibbs sampling, auxiliary variable MCMC, gradient-based MCMC)
   + importance sampling
   + SMC
+ hybrids
  + MCMC-VI
  + stochastic gradient MCMC
  + VI with local Gibbs steps
+ misc
  + hierarchical softmax
  + contrastive divergence and negative sampling
  + inducing variables
  + case-control likelihood
  + offline/compile-time algorithms, e.g., Paige and Wood (2016)

inference-independent features
+ automated constrained to unconstrained transformations
+ simulated annealing
+ simulated tempering
+ higher-order optimization methods

variational models
+ mean-field
+ structured, e.g., low rank Gaussian, normalizing flows, mixture distribution
+ external parameterizations, e.g., inference networks, classical parameter tying as in stochastic/averaged EP
+ hierarchical models
+ intractable (or costly) density, e.g. the implicit variational distribution in generative adversarial networks, MCMC, and our own recent work
+ truncated families
+ infinite families with only a finite set of distributions not set to the prior

variance reduction techniques
+ analytic KL/entropy
+ control variates
+ markov blankets/rao-blackwellization
+ local expectation gradients

This is cool. I can kind of already see the skeleton. What am I forgetting?

**Comment by datnamer** [comment]
Some ideas here for modular pieces/transforms/concepts that span stats/ml/dl : https://github.com/JuliaML/Roadmap.jl/issues/8

**Comment by dustinvtran** [comment]
Looks useful! Thanks for the link. Certainly interesting to see how others aim to organize all learning concepts.

I think we're able to support most of that roadmap (or TensorFlow as the base already does), using the core abstractions in Box's loop: `model`, `inference`, and `criticism`. I think using `Transformation` as the primary abstraction is misguided. It's a very supervised way of thinking about the world. But it could just be because I strongly believe, e.g., BDA is the best modus operandi.

**Comment by datnamer** [comment]
I agree but It seems to me that transformation is at a lower level of abstraction than model, inference and criticism  therefore they aren't mutually exclusively......though I could be wrong.

**Comment by dustinvtran** [comment]
renamed in #344
